rj_gameplay/stp/coordinator.py

```diff
@@ -0,0 +1,56 @@
+"""This module contains the implementation of the coordinator."""
+from typing import Optional
+
+import stp.play
+import stp.rc as rc
+import stp.role.assignment as assignment
+import stp.situation
+
+
+class Coordinator:
+    """The coordinator is responsible for using SituationAnalyzer to select the best
+    play to run, calling tick() on the play to get the list of actions, then ticking
+    all of the resulting actions."""
+
+    __slots__ = [
+        "_play_registry",
+        "_play_selector",
+        "_prev_situation",
+        "_prev_play",
+        "_prev_role_results",
+    ]
+
+    _play_selector: stp.situation.IPlaySelector
+    _prev_situation: Optional[stp.situation.ISituation]
+    _prev_play: Optional[stp.play.IPlay]
+    _prev_role_results: assignment.FlatRoleResults
+
+    def __init__(self, play_selector: stp.situation.IPlaySelector):
+        self._play_selector = play_selector
+
+        self._prev_situation = None
+        self._prev_play = None
+        self._prev_role_results = {}
+
+    def tick(self, world_state: rc.WorldState) -> None:
+        """Performs 1 ticks of the STP system:
+            1. Selects the best play to run given the passed in world state.
+            2. Ticks the best play, collecting the list of actions to run.
+            3. Ticks the list of actions.
+        :param world_state: The current state of the world.
+        """
+
+        # Call situational analysis to see which play should be running.
+        cur_situation, cur_play = self._play_selector.select(world_state)
```

Comment:
This will create a new play instance every tick, which kind of breaks everything. The current logic is, if the situation changes and `is_applicable` no longer applies, you can call `play.try_preempt()` which will either end the play or refuse to give up control (most plays just give up).

https://github.com/RoboJackets/robocup-software/blob/73c92878baf960844b5a4b34c72804093f1ea459/soccer/gameplay/situational_play_selection.py#L340-L348

Reply:
Resolved offline; we're going to try this stateless version to see how it works and if it is bad we'll remove it. For now, no changes needed.